lib/ansible/modules/blockinfile.py

```diff
@@ -192,14 +199,15 @@
 import re
 import os
 import tempfile
+
 from ansible.module_utils.basic import AnsibleModule
-from ansible.module_utils.common.text.converters import to_bytes, to_native
+from ansible.module_utils.common.text.converters import to_native
 
 
-def write_changes(module, contents, path):
+def write_changes(module, contents, path, encoding=None):
 
     tmpfd, tmpfile = tempfile.mkstemp(dir=module.tmpdir)
-    with os.fdopen(tmpfd, 'wb') as tf:
+    with os.fdopen(tmpfd, 'w', encoding=encoding) as tf:
```

Comment:
I had to look into this a bit, so I feel it might be worth a comment to avoid someone changing it later, but it does seem that `newlines` was explicitly omitted here to maintain the same behavior as before, so that all newlines are normalized to the system default.

However, this differs from the `open` for reading below, so a comment here might be worthwhile.

Reply:
Would it be simpler to just define it with the default value? i.e. 
```suggestion
    with os.fdopen(tmpfd, 'w', encoding=encoding, newline=None) as tf:
```


from the [python doc](https://docs.python.org/3/library/functions.html#open):
> open(file, mode='r', buffering=-1, encoding=None, errors=None, newline=None, closefd=True, opener=None)

Reply:
Happy to add a comment, maybe something like:
```suggestion
    # contents is expected to already have the desired line endings
    with os.fdopen(tmpfd, 'w', encoding=encoding) as tf:
```

Or even both:
```suggestion
    # contents is expected to already have the desired line endings
    with os.fdopen(tmpfd, 'w', encoding=encoding, newline=None) as tf:
```

Reply:
Either way, it probably still needs an explainer, so no one attempts to "harmonize" them

Reply:
^ that comment I've made is incorrect, `newline=None` translates all newline sequences (`\r`, `\r\n`, `\n`) into `\n`

Reply:
I added comments in those two places, let me know if there's any improvements you can think of